sql tuning: add explain walkthough (#3861)
pingcap/docs: explain-walkthrough.md

---
title: EXPLAIN Walkthrough 
summary: Learn how to use `EXPLAIN` by walking through an example statement
---

# EXPLAIN Walkthrough

> Because SQL is a declarative language, it is not possible to sight-check a query and tell if it is executing efficiently. We must first use [`EXPLAIN`](/sql-statements/sql-statement-explain.md) to understand what the current execution plan is.

Author: This changes the meaning. You can manually check a query (EXPLAIN). You can't automatically sight-check it.

| StreamAgg_20                 | 1.00     | root      |               | funcs:count(Column#13)->Column#11                                                                                      |
| └─TableReader_21             | 1.00     | root      |               | data:StreamAgg_9                                                                                                       |
|   └─StreamAgg_9              | 1.00     | cop[tikv] |               | funcs:count(1)->Column#13                                                                                              |
|     └─Selection_19           | 250.00   | cop[tikv] |               | ge(bikeshare.trips.start_date, 2017-07-01 00:00:00.000000), le(bikeshare.trips.start_date, 2017-07-01 23:59:59.000000) |
|       └─TableFullScan_18     | 10000.00 | cop[tikv] | table:trips   | keep order:false, stats:pseudo                                                                                         |
+------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------+
5 rows in set (0.00 sec)
```

> Working from the child operator `└─TableFullScan_18` back, you can see its execution process as follows, which is currently suboptimal:

Reviewer: ```suggestion
From the child operator `└─TableFullScan_18` back, you can see its execution process as follows, which is currently suboptimal:
```

1. The coprocessor (TiKV) reads the entire `trips` table as a `TableFullScan` operation. It then passes the rows that it reads to the `Selection_19` operator, which is still within TiKV.

> 2. The `WHERE start_date BETWEEN ..` predicate is then filtered in the `Selection_19` operator. Approximately `250` rows are estimated to meet this selection. Note that this number was produced from a heuristic; the `└─TableFullScan_18` operator shows `stats:pseudo`. After running `ANALYZE TABLE trips` the statistics should be more accurate.

Reviewer: ```suggestion
2. The `WHERE start_date BETWEEN ..` predicate is then filtered in the `Selection_19` operator. Approximately `250` rows are estimated to meet this selection. Note that this number was produced from a heuristic; the `└─TableFullScan_18` operator shows `stats:pseudo`. After running `ANALYZE TABLE trips`, the statistics should be more accurate.
```

> 3. The rows that met the selection criteria then have a `count` function applied to them. This is also completed inside the `StreamAgg_9` operator, which is still inside TiKV (`cop[tikv]`). The TiKV coprocessor understands a number of MySQL built-in functions, `count` being one of them. It also understands that `count` is safe to apply Stream Aggregation to, even though the results are not in order.

Author: I can link to aggregation once https://github.com/pingcap/docs/pull/3926 merges. Stream aggregation uses less memory than hashagg, but has rules for if it can be used.

|       └─TableFullScan_18     | 10000.00 | 19117643 | cop[tikv] | table:trips   | proc max:612ms, min:8ms, p80:248ms, p95:460ms, iters:18695, tasks:56                                                                                                                                                                              | keep order:false, stats:pseudo                                                                                         | N/A       | N/A  |
+------------------------------+----------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+-----------+------+
5 rows in set (1.03 sec)
```

The example query takes `1.03` seconds to execute. This is our benchmark to beat!

> Importantly, we can also see some of the useful columns that `EXPLAIN ANALYZE` has added. `actRows` shows that some of our estimates were off. We knew this already because the `└─TableFullScan_18` showed `stats:pseudo`. But expecting 10 thousand rows and finding 19 million rows does not show that the estimate was very accurate. If we run `ANALYZE TABLE`, and then `EXPLAIN ANALYZE` again we can see that the estimates are much closer:

Author: Sounds good to me. I've added a few links to statement-reference pages.

+------------------------------+-------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+-----------+------+
5 rows in set (0.93 sec)
```

After [`ANALYZE TABLE`](/sql-statements/sql-statement-analyze-table.md) is executed, we can see that the estimated rows for the `└─TableFullScan_18` is accurate, and the estimate for `└─Selection_19` is now also much closer. While in this case the execution plan (i.e. the set of operators TiDB uses to execute this query) has not changed in between running [`ANALYZE TABLE`](/sql-statements/sql-statement-analyze-table.md), quite frequently sub-optimal plans are caused by out of date statistics.

> As well as [`ANALYZE TABLE`](/sql-statements/sql-statement-analyze-table.md), TiDB will automatically regenerate statistics as a background operation after a [threshold is reached](/system-variables.md#tidb_auto_analyze_ratio). You can see how close TiDB is to this threshold (i.e. how healthy it considers the statistics to be) with the statement [`SHOW STATS_HEALTHY`](/sql-statements/sql-statement-show-stats-healthy.md):

Reviewer: ```suggestion
In addition to [`ANALYZE TABLE`](/sql-statements/sql-statement-analyze-table.md), TiDB automatically regenerates statistics as a background operation after a [threshold is reached](/system-variables.md#tidb_auto_analyze_ratio). You can see how close TiDB is to this threshold (how healthy TiDB considers the statistics to be) by executing the [`SHOW STATS_HEALTHY`](/sql-statements/sql-statement-show-stats-healthy.md) statement:
```

{{< copyable "sql" >}}

```sql
SHOW STATS_HEALTHY;
```

```sql
+-----------+------------+----------------+---------+
| Db_name   | Table_name | Partition_name | Healthy |
+-----------+------------+----------------+---------+
| bikeshare | trips      |                |     100 |
+-----------+------------+----------------+---------+
1 row in set (0.00 sec)
```

> ## Identifying optimizations

Reviewer: ```suggestion
## Identify optimizations
```

> There are some aspects of the current execution plan that are quite efficient:

Reviewer: ```suggestion
The current execution plan is efficient in the following aspects:
```

> * Most of the work is handled inside the TiKV coprocessor. Only 56 rows need to be sent across the network back to TiDB for processing. Each of these rows is short, and contains only the count that matched the selection.

Reviewer: ```suggestion
* Most of the work is handled inside the TiKV coprocessor. Only 56 rows need to be sent across the network back to TiDB for processing. Each of these rows is short and contains only the count that matches the selection.
```

> * Both aggregating the count of rows in TiDB (`StreamAgg_20`), and aggregating in TiKV (`└─StreamAgg_9`) are using stream aggregation, which is very efficient in its memory usage.

Reviewer: ```suggestion
* Aggregating the count of rows both in TiDB (`StreamAgg_20`) and in TiKV (`└─StreamAgg_9`) uses the stream aggregation, which is very efficient in its memory usage.
```

> The biggest issue with the current execution plan, is that the predicate `start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59'` does not apply immediately. All rows are read first with a TableFullScan operator, and then a selection is applied afterwards. If we look at the output from `SHOW CREATE TABLE trips` we can see why this is:

Reviewer: ```suggestion
The biggest issue with the current execution plan is that the predicate `start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59'` does not apply immediately. All rows are read first with a `TableFullScan` operator, and then a selection is applied afterwards. You can find out the cause from the output of `SHOW CREATE TABLE trips`:
```

{{< copyable "sql" >}}

```sql
SHOW CREATE TABLE trips\G
```

```sql
*************************** 1. row ***************************
       Table: trips
Create Table: CREATE TABLE `trips` (
  `trip_id` bigint(20) NOT NULL AUTO_INCREMENT,
  `duration` int(11) NOT NULL,
  `start_date` datetime DEFAULT NULL,
  `end_date` datetime DEFAULT NULL,
  `start_station_number` int(11) DEFAULT NULL,
  `start_station` varchar(255) DEFAULT NULL,
  `end_station_number` int(11) DEFAULT NULL,
  `end_station` varchar(255) DEFAULT NULL,
  `bike_number` varchar(255) DEFAULT NULL,
  `member_type` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`trip_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=20477318
1 row in set (0.00 sec)
```

> There is no index on `start_date`! We would need an index in order to push this predicate into an index reader operator. Let's go ahead and add one:

Reviewer: What a passion! 😀
```suggestion
There is **NO** index on `start_date`. You would need an index in order to push this predicate into an index reader operator. Add an index as follows:
```

{{< copyable "sql" >}}

```sql
ALTER TABLE trips ADD INDEX (start_date);
```

```sql
Query OK, 0 rows affected (2 min 10.23 sec)
```

> > **Tip:**

Reviewer: ```suggestion
> **Note:**
```

> 

> > The progress of DDL jobs can be monitored with the command [`ADMIN SHOW DDL JOBS`](/sql-statements/sql-statement-admin.md). The defaults in TiDB are carefully chosen so that adding an index does not impact production workloads too much. For testing environments, consider increasing [`tidb_ddl_reorg_batch_size`](/system-variables.md#tidb_ddl_reorg_batch_size) and [`tidb_ddl_reorg_worker_cnt`](/system-variables.md#tidb_ddl_reorg_worker_cnt). On a reference system, a batch size of 10240 and worker count of 32 achieved a 10x improvement over the defaults.

Reviewer: Does the `defaults` mean batch size and work count?
```suggestion
> You can monitor the progress of DDL jobs using the [`ADMIN SHOW DDL JOBS`](/sql-statements/sql-statement-admin.md) command. The defaults in TiDB are carefully chosen so that adding an index does not impact production workloads too much. For testing environments, consider increasing the [`tidb_ddl_reorg_batch_size`](/system-variables.md#tidb_ddl_reorg_batch_size) and [`tidb_ddl_reorg_worker_cnt`](/system-variables.md#tidb_ddl_reorg_worker_cnt) values. On a reference system, a batch size of 10240 and worker count of 32 can achieve a 10x performance improvement over the defaults.
```

Author: Yes

> With the index added we can then repeat the query in `EXPLAIN`. In the following output, we can see that a new execution plan is chosen, and the `TableFullScan` and `Selection` operators have been eliminated:

Reviewer: ```suggestion
After adding an index, you can then repeat the query in `EXPLAIN`. In the following output, you can see that a new execution plan is chosen, and the `TableFullScan` and `Selection` operators have been eliminated:
```

| id                          | estRows | task      | access object                             | operator info                                                     |
+-----------------------------+---------+-----------+-------------------------------------------+-------------------------------------------------------------------+
| StreamAgg_17                | 1.00    | root      |                                           | funcs:count(Column#13)->Column#11                                 |
| └─IndexReader_18            | 1.00    | root      |                                           | index:StreamAgg_9                                                 |
|   └─StreamAgg_9             | 1.00    | cop[tikv] |                                           | funcs:count(1)->Column#13                                         |
|     └─IndexRangeScan_16     | 8471.88 | cop[tikv] | table:trips, index:start_date(start_date) | range:[2017-07-01 00:00:00,2017-07-01 23:59:59], keep order:false |
+-----------------------------+---------+-----------+-------------------------------------------+-------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

> To compare the actual execution time, we can again use [`EXPLAIN ANALYZE`](/sql-statements/sql-statement-explain-analyze.md):

Reviewer: ```suggestion
To compare the actual execution time, you can again use [`EXPLAIN ANALYZE`](/sql-statements/sql-statement-explain-analyze.md):
```

|   └─StreamAgg_9             | 1.00    | 1       | cop[tikv] |                                           | time:4ms, loops:12                                                                                               | funcs:count(1)->Column#13                                         | N/A       | N/A  |
|     └─IndexRangeScan_16     | 8471.88 | 11409   | cop[tikv] | table:trips, index:start_date(start_date) | time:4ms, loops:12                                                                                               | range:[2017-07-01 00:00:00,2017-07-01 23:59:59], keep order:false | N/A       | N/A  |
+-----------------------------+---------+---------+-----------+-------------------------------------------+------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+-----------+------+
4 rows in set (0.00 sec)
```

> The query time has reduced from 1.03 seconds to 0.0 seconds. Not bad!

Reviewer: ```suggestion
From the result above, the query time has reduced from 1.03 seconds to 0.0 seconds.
```

> > **Tip:**

Reviewer: ```suggestion
> **Note:**
```

Author: I think this is a tip not a note, since it is tangential to the article.

Reviewer: Only `Note` can be correctly parsed by the docs page builder as a note format. For example: 
![image](https://user-images.githubusercontent.com/41534398/94092956-993c6e00-fe4e-11ea-92a5-7f1f3085c12d.png)

>

> > Another optimization that applies here is the coprocessor cache. If you are unable to add indexes, consider enabling the [coprocessor cache](/coprocessor-cache.md). When enabled, as long as the region has not been modified since the operator was last executed, TiKV will return the value from cache. This will also help reduce much of the expense of the expensive `TableFullScan` and `Selection` operators.

Reviewer: ```suggestion
> Another optimization that applies here is the coprocessor cache. If you are unable to add indexes, consider enabling the [coprocessor cache](/coprocessor-cache.md). When it is enabled, as long as the Region has not been modified since the operator is last executed, TiKV will return the value from the cache. This will also help reduce much of the cost of the expensive `TableFullScan` and `Selection` operators.
```
